chain/sui/builder_staking.go

```diff
@@ -0,0 +1,257 @@
+			TypeArguments: []bcs.TypeTag{},
+			Arguments: []bcs.Argument{
+				systemStateArg,
+				cointSplitResult,
+				validatorArg,
+			},
+		},
+	})
+
+	// update transaction with stake commands
+	txBase.SetInputsAndCommands(cmd_inputs, commands)
+
+	xcTx := &Tx{
+		Tx:         txBase.base,
+		public_key: fromPubkey,
+	}
+	return xcTx, nil
+}
+
+func (txBuilder TxBuilder) Unstake(args xcbuilder.StakeArgs, input xc.UnstakeTxInput) (xc.Tx, error) {
+	unstakeInput, ok := input.(*UnstakingInput)
+	if !ok {
+		return &Tx{}, errors.New("xc.StakeTxInput is not from a sui chain")
+	}
+
+	amount := args.GetAmount()
+	feePayer, ok := args.GetFeePayer()
+	if !ok {
+		feePayer = args.GetFrom()
+	}
+	fromPubkey, ok := args.GetPublicKey()
+	if !ok {
+		return nil, errors.New("sui transactions require pubkey")
+	}
+
+	txBase, err := txBuilder.newTransactionBase(
+		feePayer,
+		args.GetFrom(),
+		amount,
+		&unstakeInput.TxInput,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create transaction base: %w", err)
+	}
+
+	commands := make([]bcs.Command, 0)
+	cmd_inputs := make([]bcs.CallArg, 0)
+
+	// prepare system state input - it will be reused 'request_withdraw_stake' commands
+	systemStateInput := ArgumentInput(uint16(len(cmd_inputs)))
+	systemState, err := HexToObjectID("0x0000000000000000000000000000000000000000000000000000000000000005")
+	if err != nil {
+		return nil, fmt.Errorf("failed to encode sui system state: %w", err)
+	}
+	cmd_inputs = append(cmd_inputs, &bcs.CallArg__Object{
+		Value: &bcs.ObjectArg__SharedObject{
+			Id:                   systemState,
+			InitialSharedVersion: 1,
+			Mutable:              true,
+		},
+	})
+	packageId, err := HexToObjectID("0x0000000000000000000000000000000000000000000000000000000000000003")
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode objectId: %w", err)
+	}
+
+	// prepare unstake commands
+	for _, s := range unstakeInput.StakesToUnstake {
+		stakeIdInput := ArgumentInput(uint16(len(cmd_inputs)))
+		stakeId, err := HexToObjectID(s.ObjectId)
+		if err != nil {
+			return nil, fmt.Errorf("failed to encode stake id object: %w", err)
+		}
+		stakeDigest, err := Base58ToObjectDigest(s.Digest)
+		if err != nil {
+			return nil, fmt.Errorf("failed to encode unstake digest: %w", err)
+		}
+		cmd_inputs = append(cmd_inputs, &bcs.CallArg__Object{
+			Value: &bcs.ObjectArg__ImmOrOwnedObject{
+				Field0: stakeId,
+				Field1: bcs.SequenceNumber(s.Version),
+				Field2: stakeDigest,
+			},
+		})
+
+		commands = append(commands, &bcs.Command__MoveCall{
+			Value: bcs.ProgrammableMoveCall{
+				Package:       packageId,
+				Module:        "sui_system",
+				Function:      "request_withdraw_stake",
+				TypeArguments: []bcs.TypeTag{},
+				Arguments: []bcs.Argument{
+					systemStateInput,
+					stakeIdInput,
+				},
+			},
+		})
+	}
+
+	// check if we have to prepare a split operation
+	if unstakeInput.StakeToSplit.GetBalance().Uint64() != 0 {
+		if unstakeInput.SplitRemainder.IsZero() {
```

Comment:
The transaction input should be considered untrusted.  It would be better if the input provided the list of all necessary staking inputs, and then the trusted transaction builder contained the logic to serialize accordingly.

I.e. the user inputs `args.GetAmount()`, but here it's the remote client effectively telling the builder how much to unstake.  Not as trustworthy.

Reply:
Another alternative is we can keep the client input, but the builder should verify that the resulting amount matches the user input `args.GetAmount()`.